rfcs/0001-subgraph-composition.md

```diff
@@ -0,0 +1,233 @@
+developer would then write the DAO subgraph schema as follows:
+
+```graphql
+type _Schema_
+  @import(
+    types: ["Address", { name: "Transaction", as: "EthereumTransaction" }],
+    from: { name: "ethereum/mainnet" }
+  )
+
+type Proposal @entity {
+  id: ID!
+  proposer: Address!
+  transaction: EthereumTransaction!
+}
+```
+
+This would then allow queries that follow the references to addresses and
+transactions, like
+
+```graphql
+{
+  proposals { 
+    proposer { balance address }
+    transaction { hash block { number } }
+  }
+}
+```
+
+### Extending types from foreign schemas
+
+Extending types from another subgraph involves a few steps:
+
+1. Importing the foreign entity types.
+2. Extending these entity types with custom fields.
+3. Managing (e.g. creating) extended entities in subgraph mappings.
+
+Types can be imported either from a subgraph name or from a subgraph ID.
+Importing from a subgraph name means that the exact version of the imported
+subgraph will be identified at query time and its schema may change in arbitrary
+ways over time. Importing from a subgraph ID guarantees that the schema will
+never change but also means that the import points to a subgraph version that
+may become outdated over time.
+
+Let's say the DAO subgraph wants to extend the Ethereum `Address` type to
+include the proposals created by each respective account. To achieve this, the
+developer would write the following schema:
+
+```graphql
+type _Schema_
+  @import(
+    types: ["Address"],
+    from: { name: "ethereum/mainnet" }
+  )
+
+type Proposal @entity {
+  id: ID!
+  proposer: Address!
+}
+
+extend type Address {
+  proposals: [Proposal!]! @derivedFrom(field: "proposal")
+}
+```
+
+This makes queries like the following possible, where you can go "back" from
+accounts and contracts to proposal entities, despite the Ethereum `Address` type
+originally being defined in the `ethereum/mainnet` subgraph.
+
+```graphql
+{
+  accounts {
+    address
+    proposals {
+      id
+      proposer {
+        address
+    }
+  }
+}
+```
+
+In the above case, the `proposals` field on the extended type is derived, which
+means that an implementation likely wouldn't have to create a local extension
+type in the store. However, if `proposals` was defined as
+
+```graphql
+extend type Address {
+  proposals: [Proposal!]!
+}
+```
+
+then it would have to be possible for the subgraph mappings to create `Address`
+instances and set the `proposals` field. These entity instances would have to be
+merged with the corresponding original entities from the `ethereum/mainnet`
+subgraph.
+
+### Subgraph availability
+
+In the decentralized network, queries will be split and routed through the
+network based on what indexers are available and which subgraphs they index.
+At that point, failure to find an indexer for a subgraph that types were
+imported from will result in a query error.
+
+Until the network is reality, we are dealing with individual Graph nodes and
```

Comment:
What about in the case where the foreign type is not nullable in the GraphQL schema?

Reply:
This same open question exists when mixing different chains and networks in a same subgraph, the entities need to be merged by id and a mandatory field may be missing.

IMO tolerance is the right guiding principle, and a way to be tolerant here but not violate GraphQL rules is to omit entities which don't have all their mandatory fields after merging from the result.

Reply:
Are we talking about
```
type SomeEntity @entity {
  foo: ForeignType!
}
```
where the subgraph from which `ForeignType` was imported is not available and the `foo` value would be `null`?

I don't think we can exclude entities that exist but are not broken from query results. IMHO what should happen in the above case is that the GraphQL query is satisfied but there is an error for the branch of the query that involves the entity. We should compare to the error behavior other GraphQL servers (e.g. ApolloServer) have when a non-nullable field in a response is not set.

Reply:
The graphql spec has discussion on this exact case https://graphql.github.io/graphql-spec/draft/#sec-Errors.Error-result-format.

> If the field which experienced an error was declared as Non-Null, the null result will bubble up to the next nullable field.

This is similar to our suggestion, but the entity can't just vanish, it should be replaced by `null`.

If we include errors with data in our protocol, we need to consider what that will look like for the read interface and attestations.

Reply:
@leoyvens Yep, that's a good point. I think we _do_ want to include errors like this in the response, so they will have to be part of the read interface / attestations as well. The query engine should either pick a node that has both/all subgraphs involved, or it will split the query up at the subgraph boundaries (which I'm less sure about in the composition case than in regular queries), but even then, the data the query engine has may be out of date or the index node may remove one of the subgraphs involved before it processes the query.

Everything but an easy situation _but_ for now I'll include some details about this paticular case with the desired error behavior to the RFC.